### ISSUE TITLE
Linux ARM64 build

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#3515]
https://forums.zotero.org/discussion/109836/add-support-for-asahi-linux-arm-apple-hardware

Supposedly Mozilla [has (or had) untested ARM64 builds in CI](https://bugzilla.mozilla.org/show_bug.cgi?id=1678342#c12). I'm not sure where those are, but if we can find them we can try a build.

**Comment by dstillman** [comment]
(I mean, we could also just make an ARM64 build, but I'd rather not do that, particularly on Linux where we don't even need to use a custom launcher at the moment.)

**Comment by amazingfate** [comment]
Archlinuxcn has packaged zotero for arm64: https://github.com/archlinuxcn/repo/blob/master/alarmcn/zotero/PKGBUILD.
But the current build system for Linux sucks: it still downloads Firefox binaries from somewhere, which prevents it from get packaged by distros like debian. And zotero is depending on Firefox v60, which is a really old release from 2018.

**Comment by amazingfate** [comment]
You can use firefox-esr from ppa: https://launchpad.net/~mozillateam/+archive/ubuntu/ppa/+packages?field.name_filter=firefox-esr&field.status_filter=&field.series_filter=
But I still don't think it's a good idea to download binaries at build time. If zotero can depend on a large range of Firefox versions we can just copy Firefox binaries from distro packages installed on system.

**Comment by dstillman** [comment]
> If zotero can depend on a large range of Firefox versions

It can't. It's literally a specific version of Firefox, with changes layered on top.

And we don't produce an Ubuntu-specific version — we produce a generic Linux tarball, based on the Firefox tarball, that needs to be portable to most Linux systems. I'm not sure how PPA builds are built, but if they're built with dependencies on specific Ubuntu versions, that wouldn't work.

**Comment by dstillman** [comment]
> It's impossible to build one Firefox binary which can run all Linux versions of distributions.

Right, but the tarball from Mozilla is obviously designed to be as portable as possible. So the question is whether the PPA build has appreciably different [minimum library versions](https://www.mozilla.org/en-US/firefox/115.0/system-requirements/#software-requirements) or other dependencies. But it sounds like you're saying it doesn't. A 2018 cut-off would be fine with us.

If we resolve the problems on the npm side of things, we could try a build from that, though we'd first need to create a script that massaged the deb files into something that exactly mirrored the Mozilla tarball. I'd still be interested in finding Mozilla's own CI builds, which would be easier to integrate.

We'd also need to build a custom updater for ARM64, which the .deb obviously doesn't have.

**Comment by dstillman** [comment]
If this is something you're interested in exploring, and you can get `npm run build` to work (even if it requires removing some `reader` dependencies for now), we'd ultimately want `fetch_deb` and `deb2tarball` scripts in `app/linux/` that took a Firefox version number (as in `GECKO_VERSION_LINUX` from `app/config.sh`) as a command-line parameter. The first would download the latest published .deb for that Firefox version and save it to `app/xulrunner/` with a name based on the version, and the second would convert the .deb file in `app/xulrunner/` to a .tar.gz with the expected Mozilla layout, symlinks removed, etc. We would then add support in `app/scripts/fetch_xulrunner` to call those two scripts for aarch64 when making a Linux build, and the ARM64 build would be built automatically along with the i686 and x86_64 builds when calling `app/scripts/dir_build -p l`.

**Comment by amazingfate** [comment]
> > It's impossible to build one Firefox binary which can run all Linux versions of distributions.
> 
> Right, but the tarball from Mozilla is obviously designed to be as portable as possible. So the question is whether the PPA build has appreciably different [minimum library versions](https://www.mozilla.org/en-US/firefox/115.0/system-requirements/#software-requirements) or other dependencies. But it sounds like you're saying it doesn't. A 2018 cut-off would be fine with us.

firefox v115 need libxtst 1.2.3, which is packaged in ubuntu since ubuntu 18.04: https://launchpad.net/ubuntu/+source/libxtst

**Comment by bu6n** [comment]
I managed to use a fedora build of firefox 102.0 to make it compile (on asahi linux with a macbook M1 pro), but when the app starts (`./app/staging/Zotero_linux-aarch64/zotero`) it does only seem to load a blank firefox page. @amazingfate do you know where I could find a build of firefox 102.15.1esr? @dstillman you mention a work in progress with firefox 115, should that already work on the `main` branch?

**Comment by amazingfate** [comment]
> You can use firefox-esr from ppa: https://launchpad.net/~mozillateam/+archive/ubuntu/ppa/+packages?field.name_filter=firefox-esr&field.status_filter=&field.series_filter=

@bu6n

**Comment by bu6n** [comment]
I can't seem to find a find for 102.15.1, only references to builds?

**Comment by amazingfate** [comment]
I guess 102.15.0 should be okay.

**Comment by bu6n** [comment]
They don't have linked files there?

**Comment by bu6n** [comment]
@amazingfate Would you be able to test the code I pushed in the PR?

**Comment by ayakael** [comment]
I'm working on getting `aarch64` working on my Alpine Linux zotero package. (https://lab.ilot.io/ayakael/user-aports/-/jobs/8756) I have those issues with phantomjs. Can anyone elaborate on how I should go about fixing it? Many thanks!

**Comment by bu6n** [comment]
I'm on Fedora 39 and haven't encountered this particular issue. I've built the master branch if that makes a difference?

**Comment by ayakael** [comment]
I've fixed the issue that I had by dropping phantomjs as it isn't even used: https://github.com/zotero/epub.js/pull/3

**Comment by fredrikrab** [comment]
Firefox Nightly is now available for Linux ARM64.

Blog post: https://blog.nightly.mozilla.org/2024/04/19/firefox-nightly-now-available-for-linux-on-arm64/

> We want to be upfront about the current state of our ARM64 builds. Although we are confident in the quality of Firefox on this architecture, we are still incorporating comprehensive ARM64 testing into Firefox’s continuous integration and release pipeline. Our goal is to integrate ARM64 builds into Firefox’s extensive automated test suite, which will enable us to offer this architecture across the beta, release, and ESR channels.

Download: https://www.mozilla.org/en-US/firefox/all/#product-desktop-nightly

**Comment by dstillman** [comment]
Once there are ARM64 Linux tarballs for Firefox ESR (and we've built a custom updater), we'll be able to offer this.

**Comment by thomas-bouvier** [comment]
Nightly tarballs are now available for Firefox https://www.mozilla.org/en-US/firefox/all/desktop-nightly/. Let's wait for ESR :)

**Comment by mkurz** [comment]
@thomas-bouvier Yes, but this is nothing new or not? That was pointed out in May already in [a previous comment.](https://github.com/zotero/zotero/issues/3515#issuecomment-2126266809)

**Comment by utopianreflection** [comment]
While we wait for a Firefox ESR on aarch64-linux, I have created and (slightly) tested a Nix flake which builds Zotero on aarch64-linux (with some slight patches, so that it could be Nixified):

https://gitlab.com/utopianreflection/zotero-flake

This flake also has a x86_64-linux output for people who would like to include Zotero in their configurations, but who want to build it from source or build a patched version.

Let me know if this is helpful or any issues/optimizations you find.

**Comment by ayakael** [comment]
Of note, starting with Alpine 3.21, due for release in a few weeks, Zotero is available on x86_64, x86, aarch64 and armv7. Abuild for 3.20 is available [here](https://ayakael.net/forge/-/packages/alpine/zotero/7.0.9-r0). The package is a built-from-source Firefox patched for Alpine, and Zotero is them built against it.

**Comment by Serjicalclic** [comment]
Noob here, but it seems Debian released a version of Firefox ESR 128.4.0 for arm64 : [https://packages.debian.org/sid/arm64/firefox-esr/download](https://packages.debian.org/sid/arm64/firefox-esr/download)
Would it be enough, @dstillman ?

**Comment by dstillman** [comment]
No. Zotero isn't built exclusively for Debian. We need an official Mozilla build.

(We're also not yet built on Firefox 128 ESR, so if this becomes available for 128, as is likely, we'll need to update first if we haven't yet.)

**Comment by jace** [comment]
> Of note, starting with Alpine 3.21, due for release in a few weeks, Zotero is available on x86_64, x86, aarch64 and armv7. Abuild for 3.20 is available [here](https://ayakael.net/forge/-/packages/alpine/zotero/7.0.9-r0). The package is a built-from-source Firefox patched for Alpine, and Zotero is then built against it.

@ayakael how do I use this? I'm new to Alpine Linux. I've installed Alpine in a proot in Termux (aarch64) and setup the repository, but it has unmet dependencies.

```
localhost:~# apk add zotero
ERROR: unable to select packages:
  ffmpeg-libavcodec (no such package):
    required by: zotero-7.0.10-r0[ffmpeg-libavcodec]
  so:libevent-2.1.so.7 (no such package):
    required by: zotero-7.0.10-r0[so:libevent-2.1.so.7]
  so:libffi.so.8 (no such package):
    required by: zotero-7.0.10-r0[so:libffi.so.8]
  so:libicui18n.so.74 (no such package):
    required by: zotero-7.0.10-r0[so:libicui18n.so.74]
  so:libicuuc.so.74 (no such package):
    required by: zotero-7.0.10-r0[so:libicuuc.so.74]
  so:libvpx.so.9 (no such package):
    required by: zotero-7.0.10-r0[so:libvpx.so.9]
```

I guess I need to install some other repositories? Which ones?

**Comment by jace** [comment]
> @ayakael how do I use this? I'm new to Alpine Linux. I've installed Alpine in a proot in Termux (aarch64) and setup the repository, but it has unmet dependencies.

Resolved -- Andronix gave me Alpine 3.10. I switched to [proot-distro](https://github.com/termux/proot-distro) which defaults to `edge` and now I'm able to launch Zotero with this invocation:

```bash
pd login alpine --shared-tmp --user aid_u0_a503 --termux-home -- DISPLAY=:0.0 zotero
```

This is way too much effort to get Zotero with extended plugin-based functionality on my phone, but I'm glad I've got to a working invocation. 😅

**Comment by yuyichao** [comment]
Update from the maintainer of the aarch64 zotero build in the archlinuxcn repo here.

First a clarification, even though my previous setup uses a ubuntu build of firefox, it is sufficiently standalone (after all, mozilla also built their official binary somewhere) that it should work on other distros as well and it seems that other people have used it successfully. Also, what I did has always been irrelevant for any distro with enough resources to use a source-built firefox-esr.

I've recently updated my build scripts to work with zotero 7. The ppa I was using previously seems to stop producing a binary package for all the build it did so I couldn't use that. I tried the debian binary but it depends on external packages (libffi.so.7 to be more specific) that isn't available on arch (and probably some other new distros). That's why I decided to try the nightly build. I don't know what's the nightly version that's the closest to a particular esr version so I simply picked the latest nightly that still bear the same major version. It's using v128 since that's the first esr after the aarch64 nightly become available.

For anyone that needs it, this is the patch I made https://github.com/yuyichao/zotero/commit/0e1655e4e873e3df354a3c49d40cacb3bb79f223#diff-8de1cfc9a1fc7b986666935d2438c603cd6cc4c75963c517db9e16f18959880cR913. It's much more minimum thanks to the use of an official binary with similar package structure as the released one. I assume it's still not upstreamable but hopefully mozilla will release a esr binary for linux-aarch64 so that the platform can be officially supported.

**Comment by dstillman** [comment]
Mozilla now has ARM64 builds for Firefox 136.0b8, but the changes they've made since Firefox 128 ESR are too extensive to create a version that works in both Firefox 128 and 136, so we'll need to wait for Firefox 140 ESR this summer before offering this.

**Comment by mkurz** [comment]
> so we'll need to wait for Firefox 140 ESR this summer before offering th

Firefox 140 ESR released: https://www.mozilla.org/en-US/firefox/140.0/releasenotes/

**Comment by dstillman** [comment]
Now available in the Zotero 8 beta:

https://forums.zotero.org/discussion/125835/available-for-beta-testing-zotero-for-linux-on-arm64

**Comment by mkurz** [comment]
If you are on Arch Linux ARM you can use this AUR - works perfectly for me: https://aur.archlinux.org/packages/zotero-beta-bin
e.g. with yay: `yay -S zotero-beta-bin`
